app/src/main/java/com/hedvig/app/feature/marketing/MarketingViewModel.kt

```diff
@@ -56,108 +55,61 @@ class MarketingViewModel(
-        availableMarkets = availableMarkets,
-      )
-    }
-  }
+  private suspend fun getAvailableMarkets() = listOfNotNull(
+    Market.SE,
+    Market.NO,
+    Market.DK,
+    if (featureManager.isFeatureEnabled(Feature.FRANCE_MARKET)) {
+      Market.FR
+    } else {
+      null
+    },
+  )
 
   fun setMarket(market: Market) {
-    val newState = updateMarketPickerState { it.copy(market = market) }
-    safeLet((newState as? PickMarket)?.market, (newState as? PickMarket)?.language) { m, l ->
-      updateApplicationLanguageUseCase.invoke(m, l)
-    }
-  }
-
-  fun setLanguage(language: Language) {
-    val newState = updateMarketPickerState { it.copy(language = language) }
-    safeLet((newState as? PickMarket)?.market, (newState as? PickMarket)?.language) { m, l ->
-      updateApplicationLanguageUseCase.invoke(m, l)
-    }
-  }
+    updateApplicationLanguageUseCase.invoke(market, market.defaultLanguage())
 
-  private fun updateMarketPickerState(transform: (oldState: PickMarket) -> PickMarket): ViewState {
-    val oldState = state.value as? PickMarket
-      ?: throw IllegalStateException("Cannot update market picker state when not on market picker")
-    val newState = transform(oldState)
-    val newStateWithLanguage = newState.copy(
-      language = languageFromState(newState),
-    )
-    return _state.updateAndGet {
-      newStateWithLanguage.copy(
-        isValid = isValid(newStateWithLanguage),
+    _state.update {
+      it.copy(
+        market = market,
+        language = market.defaultLanguage(),
       )
     }
   }
 
-  private fun isValid(state: PickMarket): Boolean {
-    return (state.market != null && state.language != null)
-  }
-
-  private fun languageFromState(state: PickMarket): Language? {
-    val market = state.market ?: return state.language
-    val language = state.language ?: return market.defaultLanguage()
-
-    if (!market.isCompatible(language)) {
-      return market.defaultLanguage()
+  fun setLanguage(language: Language) {
+    state.value.market?.let {
+      updateApplicationLanguageUseCase.invoke(it, language)
     }
 
-    return language
+    _state.update { it.copy(language = language) }
   }
 
   fun submitMarketAndLanguage() {
-    val currentState = state.value as? PickMarket ?: throw IllegalStateException("Can't submit, not in PickMarket")
-    if (!currentState.isValid) {
-      throw IllegalStateException("Can't submit, PickMarket is not valid")
-    }
-    val market = currentState.market ?: throw IllegalStateException("Can't submit, market is null")
-    val language = currentState.language ?: throw IllegalStateException("Can't submit, language is null")
-    updateMarketPickerState { it.copy(isLoading = true) }
     viewModelScope.launch {
-      val result = submitMarketAndLanguagePreferencesUseCase
-        .invoke(language, market)
-      when (result) {
+      val market = _state.value.market ?: throw IllegalArgumentException("Market null")
```

Comment:
nit: I like replacing this throw.. with `?: error("Market null")` which is more concise and actually throws the semantically correct exception of `IllegalStateException` which is the case the majority of the times. IllegalArgument would fit more for arguments passed into this function, but in the current case it's the state which is in a "non correct" state for this function to be called.
Same for the line below of course.